### PR TITLE
Update the indexRandom function to create more segments

### LINKED ISSUE #10310: Concurrent search tests run with a slice count of 1
As part of https://github.com/opensearch-project/OpenSearch/issues/9443, we worked on parameterizing the tests with concurrent search enabled, such that any test which is testing the search use-case gets tested with both the paths. However, the tests that are running with concurrent search enabled feature have slice count of 1 which means that the concurrent search test path is not getting tested completely.

This issue is to fix the tests such that all the parameterized tests and concurrent search enabled run with slice count of greater than 1.

### PR DESCRIPTION
**Pull request by neetikasinghal**
### Description
This is to make sure that the index random function which is used in multiple classes has creation of multiple segments to test the concurrent search code path. Without this change there are 848 IT tests that are running with slice count<=1.

With this change the number of tests having slice<=1 is down to 325. This pr can also be used to add index random in the other tests not running with the higher slice count. e.g. https://github.com/opensearch-project/OpenSearch/commit/6157d4e0065121789a70d40c44349c3f331d262b

### Related Issues
Resolves #10310

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by neetikasinghal** [comment]
@sohami @reta please review.

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java [hunk 1]
```diff
@@ -1659,6 +1663,51 @@ public void indexRandom(boolean forceRefresh, boolean dummyDocuments, boolean ma
         }
     }
 
+    /*
+    * This method ingests bogus documents for the given indices such that multiple slices
+    * are formed. This is greatly useful for testing with the concurrent search use-case.
+    * @param refreshCount    the number of segments to be formed by ingesting bogus documents
+    * @param indices         the indices in which bogus documents should be ingested
+    * */
+    public Set<List<String>> indexRandomForMultipleSlices(int segments, String... indices) throws InterruptedException {
```
Comment (reta) [inline test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1672]: We probably shouldn't thrown `InterruptedException` but `RuntimeException` instead (it pops up in all test cases otherwise). Also, I think this method should be `protected`

File: test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java [hunk 2]
```diff
@@ -44,4 +46,14 @@ public void afterTests() {
...
     }
+
+    public void indexRandomForConcurrentSearch(int segments, String... indices) throws InterruptedException {
+        if (dynamicSettings.get(CLUSTER_CONCURRENT_SEGMENT_SEARCH_SETTING.getKey()).equals("true")) {
+            indexRandomForMultipleSlices(segments, indices);
+        }
+    }
+
+    public void indexRandomForConcurrentSearch(String... indices) throws InterruptedException {
+        indexRandomForConcurrentSearch(2, indices);
```
[the remark this answers is not in the document]
Comment (neetikasinghal) [reply test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java:57]: these functions rely on `dynamicSettings` and used for the parameterized tests only. Hence, i thought of keeping these functions in this class. thoughts?

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java [hunk 3]
```diff
@@ -1659,6 +1663,51 @@ public void indexRandom(boolean forceRefresh, boolean dummyDocuments, boolean ma
...
+        Set<List<String>> bogusIds = new HashSet<>();
+        for (String index : indices) {
+            int numDocs = getNumShards(index).totalNumShards * segments;
```
[the remark this answers is not in the document]
Comment (neetikasinghal) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1675]: all the tests have been tested with multiple of segments as 2/3 - for most of the tests multiple of 2 seems to be working. Why do we want to have randomization here? If we want this, then we will have to test out all the tests again to make sure we have minimum slice count of 2. Theoretically, it should, however it needs to be tested again.

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java (hunk 1 above)
Comment (neetikasinghal) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1672]: @reta I know what u are referring to but we have a latch usage in the function, hence its throwing `InterruptedException`. Let me know if you have any suggestions.

Also, I agree to make it protected, I will make that change.

Comment (reta) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1672]:
```
try {
 ...
} catch (InterruptedException ex) {
    Thread.currentThread().interrupt();
    throw new RuntimeException("The wait was interrupted", ex");
}
```

File: test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sohami) [reply test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java:57]: For dynamic settings, these functions can fetch it from cluster setting using `client().admin().cluster()...`. That way I see it more like a generic utility nothing specific to parameterization. That way any tests should be able to use these utility as needed.

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sohami) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1675]: the point is using `segments` variable as a multiplier is confusing to me. You can either use randomized constant or just some fixed constant like 2/3 whichever works.

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java (hunk 1 above)
Comment (neetikasinghal) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1672]: indexRandom function is also throwing `InterruptedException` exception. Hence, I was thinking to do the same.

I know if we change it to RunTimeException, we will have to change the tests calling this method. But we are anyways making changes in the tests, this shouldn't be a big adhoc. What do u think?

File: test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (neetikasinghal) [reply test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java:57]: the way to fetch cluster settings in the class is via internalCluster like - `internalCluster().clusterService().getClusterSettings()` - since internal cluster is already loaded before the parameterization, it doesn't contain the updated value of concurrent search done by the parameterization. Hence, keeping it in the parameterized class only.

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (neetikasinghal) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1675]: got it, randomizing this count between 2&3 as these values have been tested for most of the tests.

**Comment by neetikasinghal** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/28869/
> * **CommitID:** [4b53b81](https://github.com/opensearch-project/OpenSearch/commit/4b53b817f0f14661b4a597594c2ef5ee19c65393)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

fixing most of the tests by adding a check for dummyDocuments in the indexRandom function, however there are two tests that are flaky for concurrent search with slice count >=2.:

org.opensearch.search.fetch.subphase.highlight.HighlighterSearchIT.testHighlightNoMatchSizeWithMultivaluedFields {p0={"search.concurrent_segment_search.enabled":"true"}}
org.opensearch.search.fetch.subphase.highlight.HighlighterSearchIT.testHighlightNoMatchSize {p0={"search.concurrent_segment_search.enabled":"true"}}

Will create a github issue to track them.

File: test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sohami) [reply test/framework/src/main/java/org/opensearch/test/ParameterizedOpenSearchIntegTestCase.java]: I would expect parameterization to update the cluster setting before running the tests. The utility function is called from each test so it should get updated settings. You may need to use the mechanism I shared above to explicitly make the call if the method you shared is fetching the cached value for some reason. But for now I am fine with keeping it here as well.

File: test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java [hunk 4]
```diff
@@ -1683,6 +1688,52 @@ public void indexRandom(boolean forceRefresh, boolean dummyDocuments, boolean ma
...
+    * are formed. This is useful for testing with the concurrent search use-case as it creates
+    * multiple slices based on segment count.
+    * @param indices         the indices in which bogus documents should be ingested
+    * */
+    protected Set<List<String>> indexRandomForMultipleSlices(String... indices) throws InterruptedException {
+        Set<List<String>> bogusIds = new HashSet<>();
+        int refreshCount = randomIntBetween(2, 3);
+        for (String index : indices) {
+            int numDocs = getNumShards(index).totalNumShards * refreshCount;
+            while (refreshCount-- > 0) {
```
Comment (sohami) [inline test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java:1702]: this is still following the original behavior. My callout is to keep the multiplier here different than `segments/refreshCount` as those are unrelated. Any change in one can affect the other which is what we don't want.

**Comment by neetikasinghal** [comment]
> > ### Gradle Check (Jenkins) Run Completed with:
> > 
> > * **RESULT:** FAILURE ❌
> > * **URL:** https://build.ci.opensearch.org/job/gradle-check/28869/
> > * **CommitID:** [4b53b81](https://github.com/opensearch-project/OpenSearch/commit/4b53b817f0f14661b4a597594c2ef5ee19c65393)
> >   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
> >   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?
> 
> fixing most of the tests by adding a check for dummyDocuments in the indexRandom function, however there are two tests that are flaky for concurrent search with slice count >=2.:
> 
> org.opensearch.search.fetch.subphase.highlight.HighlighterSearchIT.testHighlightNoMatchSizeWithMultivaluedFields {p0={"search.concurrent_segment_search.enabled":"true"}} org.opensearch.search.fetch.subphase.highlight.HighlighterSearchIT.testHighlightNoMatchSize {p0={"search.concurrent_segment_search.enabled":"true"}}
> 
> Will create a github issue to track them.

created this https://github.com/opensearch-project/OpenSearch/issues/10900 to track these

**Comment by neetikasinghal** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/29006/
> * **CommitID:** [5ca93a4](https://github.com/opensearch-project/OpenSearch/commit/5ca93a4843a35d25155a722b0163c2318ee05740)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/10902 - flaky
https://github.com/opensearch-project/OpenSearch/issues?q=is%3Aissue+is%3Aopen+RemoteStoreStatsIT - many tests are flaky

**Comment by neetikasinghal** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/29008/
> * **CommitID:** [753008b](https://github.com/opensearch-project/OpenSearch/commit/753008bb870b600a2dc3ce63f736bc1b1cc47a58)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/10902 - flaky
https://github.com/opensearch-project/OpenSearch/issues?q=is%3Aissue+is%3Aopen+RemoteStoreStatsIT - many tests are flaky

**Review by sohami — APPROVED** [review]
@reta LGTM. Please let me know if you have any open comments.
